packages/core/src/di/injector_compatibility.ts

```diff
@@ -240,17 +211,22 @@ export function inject<T>(token: ProviderToken<T>, options: InjectOptions): T|nu
  */
 export function inject<T>(
     token: ProviderToken<T>, flags: InjectFlags|InjectOptions = InjectFlags.Default): T|null {
-  if (typeof flags !== 'number') {
-    // While TypeScript doesn't accept it without a cast, bitwise OR with false-y values in
-    // JavaScript is a no-op. We can use that for a very codesize-efficient conversion from
-    // `InjectOptions` to `InjectFlags`.
-    flags = (InternalInjectFlags.Default |  // comment to force a line break in the formatter
-             ((flags.optional && InternalInjectFlags.Optional) as number) |
-             ((flags.host && InternalInjectFlags.Host) as number) |
-             ((flags.self && InternalInjectFlags.Self) as number) |
-             ((flags.skipSelf && InternalInjectFlags.SkipSelf) as number)) as InjectFlags;
-  }
-  return ɵɵinject(token, flags);
+  return ɵɵinject(token, convertToBitFlags(flags));
+}
+
+// Converts object-based DI flags (`InjectOptions`) to bit flags (`InjectFlags`).
+export function convertToBitFlags(flags: InjectOptions|InjectFlags|undefined): InjectFlags|
```

Comment:
```suggestion
export function convertToBitFlags(flags: undefined): undefined;
export function convertToBitFlags(flags: InjectOptions|InjectFlags): InjectFlags;
export function convertToBitFlags(flags: InjectOptions|InjectFlags|undefined): InjectFlags|
```

Setting up the signature like this will allow you to skip the `as InjectFlags` cast later when TS knows that `flags` is defined.